lib/apollo/withData.js

```diff
@@ -1,85 +1,87 @@
 import React from 'react'
-import PropTypes from 'prop-types'
-import { ApolloProvider, getDataFromTree } from 'react-apollo'
-import initApollo from './initApollo'
+import initApolloClient from './initApollo'
+import Head from 'next/head'
+import { getDataFromTree } from 'react-apollo'
 
-// Gets the display name of a JSX component for dev tools
-function getComponentDisplayName (Component) {
-  return Component.displayName || Component.name || 'Unknown'
-}
+export default App => {
+  return class WithApolloClient extends React.Component {
+    static async getInitialProps (pageContext) {
+      const { Component, router } = pageContext
 
-export default ComposedComponent => {
-  class WithData extends React.Component {
-    static async getInitialProps (ctx) {
-      let serverState = {}
-      // Evaluate the composed component's getInitialProps()
-      let composedInitialProps = {}
-      if (ComposedComponent.getInitialProps) {
-        composedInitialProps = await ComposedComponent.getInitialProps(ctx)
+      let appProps = {}
+      if (App.getInitialProps) {
+        appProps = await App.getInitialProps(
+          pageContext
+        )
       }
 
-      // we forward the accept header for webp detection
-      // - never forward cookie to client!
       const headers = !process.browser
-        ? {accept: ctx.req.headers.accept}
+        ? pageContext.ctx.req.headers
```

Comment:
should probably be `pageContext.req.headers` or actually `{accept: pageContext.req.headers.accept}` to not leak cookies